src/app/persistence/AttributePersistence.cpp

```diff
@@ -0,0 +1,138 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/ConcreteAttributePath.h>
+#include <app/data-model/Nullable.h>
+#include <app/persistence/AttributePersistence.h>
+#include <app/persistence/AttributePersistenceProvider.h>
+#include <app/persistence/PascalString.h>
+
+namespace chip::app {
+
+namespace {
+
+bool VerifySuccessLogOnFailure(const ConcreteAttributePath & path, CHIP_ERROR err)
+{
+    VerifyOrReturnValue(err != CHIP_NO_ERROR, true);
+
+    // assume value not found being typical.
+    VerifyOrReturnValue(err != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND, false);
+
+    ChipLogError(Zcl, "Failed to load attribute %u/" ChipLogFormatMEI "/" ChipLogFormatMEI ": %" CHIP_ERROR_FORMAT,
+                 path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId), err.Format());
+    return false;
+}
+
+void LogInvalidPascalContent(const ConcreteAttributePath & path)
+{
+    ChipLogError(Zcl, "Invalid pascal content: %u/" ChipLogFormatMEI "/" ChipLogFormatMEI, path.mEndpointId,
+                 ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId));
+}
+
+template <typename StringType>
+bool LoadPascalStringImpl(AttributePersistenceProvider & provider, const ConcreteAttributePath & path, StringType & value,
+                          std::optional<typename StringType::ValueType> valueOnLoadFailure)
+{
+    MutableByteSpan rawBytes = value.RawFullBuffer();
+
+    if (VerifySuccessLogOnFailure(path, provider.ReadValue(path, rawBytes)))
+    {
+        // value is read, however it has to be a valid pascal value
+        // The value MUST be valid over the read range (i.e. rawBytes)
+        auto concrete = reinterpret_cast<const typename StringType::ValueType::pointer>(rawBytes.data());
+        if (value.IsValid({ concrete, rawBytes.size() }))
+        {
+            return true;
+        }
+
+        LogInvalidPascalContent(path);
+    }
+
+    // failed, try to set some default
+    if (!valueOnLoadFailure.has_value() || !value.SetValue(*valueOnLoadFailure))
+    {
+        value.SetNull();
+    }
+    return false;
+}
+
+} // namespace
+
+bool AttributePersistence::Load(const ConcreteAttributePath & path, Storage::ShortPascalString & value,
+                                std::optional<CharSpan> valueOnLoadFailure)
+{
+    return LoadPascalStringImpl(mProvider, path, value, valueOnLoadFailure);
+}
+
+bool AttributePersistence::Load(const ConcreteAttributePath & path, Storage::ShortPascalBytes & value,
+                                std::optional<ByteSpan> valueOnLoadFailure)
+{
+    return LoadPascalStringImpl(mProvider, path, value, valueOnLoadFailure);
+}
+
```

Comment:
Why do we expose the "long" versus "short" pascal strings? How many Long pascal strings are there really? This seems like layer crossing that causes odd linkage to old Ember data types that are not coherent with Matter. Could all strings be stored as "long" but we attempt a short load, and if it overflows, we fall back to trying long? This should always work going forward.

Reply:
Sure, sounds good. I would make LongPascalString return an unsupported error right now to make it clear "Hey we don't have this".